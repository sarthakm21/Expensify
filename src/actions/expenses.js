import db from "../firebase/firebase";

//ADD_EXPENSE
const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return db
      .ref("expenses") //returning this so that the tests can be run on this
      .push(expense)
      .then((data) => {
        dispatch(
          addExpense({
            id: data.key,
            ...expense,
          })
        );
      })
      .catch((e) => console.log("Some error occured", e));
  };
};

//REMOVE_EXPENSE
const removeExpense = (id) => ({
  type: "REMOVE_EXPENSE",
  id,
});

const startRemoveExpense = (id) => {
  return (dispatch) => {
    db.ref(`expenses/${id}`)
      .remove()
      .then((data) => {
        dispatch(removeExpense(id));
      })
      .catch((e) => console.log("Some error occured", e));
  };
};

//EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

const startEditExpense = (id, updates) => {
  return (dispatch) => {
    db.ref(`expenses/${id}`)
      .update(updates)
      .then((data) => {
        dispatch(editExpense(id, updates));
      })
      .catch((e) => console.log("Some error occured", e));
  };
};

//SET EXPENSE
const setExpense = (expenses) => ({
  type: "SET_EXPENSE",
  expenses,
});

const startSetExpense = () => {
  return (dispatch) => {
    return db
      .ref("expenses")
      .once("value")
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((child) => {
          expenses.push({ id: child.key, ...child.val() });
        });
        dispatch(setExpense(expenses));
      })
      .catch((e) => console.log(e));
  };
};

export {
  addExpense,
  removeExpense,
  editExpense,
  setExpense,
  startAddExpense,
  startEditExpense,
  startRemoveExpense,
  startSetExpense,
};
