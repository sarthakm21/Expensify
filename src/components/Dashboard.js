import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from "./ExpenseSummary";

const Dashboard = () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
export default Dashboard;
