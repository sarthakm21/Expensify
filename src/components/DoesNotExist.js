import React from "react";
const DoesNotExist = () => <div>404! The page requested does not exist.</div>;
export default DoesNotExist;
