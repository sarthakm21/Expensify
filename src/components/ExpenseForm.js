import React, { Component } from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends Component {
  //Using state directly without constructor because of the babel plugin "@babel/plugin-proposal-class-properties",
  //Using class properties instead of methods for the same reason
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? props.expense.amount.toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: "",
    };
  }

  //Declaring them as class properties instead of methods(by using arrow function) to prevent using 'this' binding
  /* Never use e.target.value directly in the setState because its callback is asynchronous and by the time
    it executes, it is possible that the value no longer exists! */
  handleDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  handleNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  handleAmountChange = (e) => {
    const amount = e.target.value;
    (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) &&
      this.setState(() => ({ amount }));
    //Using a regex to match favourable strings. Refer http://www.regex101.com
    console.log(this.state.amount);
  };

  handleSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please fill out both description and amount",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  // -------------------SingleDatePicker Component's event handlers ------------------------------------

  handleDateChange = (createdAt) => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };

  handleFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };

  render() {
    return (
      <div>
        {this.state.error}
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.handleDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.handleAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.handleDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.handleFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false} //To allow user to enter the dates in the past as well
            displayFormat="DD MMM YYYY"
          />
          <textarea
            placeholder="Add some note for the expense"
            value={this.state.note}
            onChange={this.handleNoteChange}
          ></textarea>
          <button>{this.props.text}</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
