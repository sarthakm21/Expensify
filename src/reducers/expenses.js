export default (state = [], action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];

    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);

    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates,
          };
        } else return expense;
        //Object spread syntax does not work with many browsers so we need to add a babel plugin for this.
      });

    case "SET_EXPENSE":
      return action.expenses;

    default:
      return state;
  }
};
