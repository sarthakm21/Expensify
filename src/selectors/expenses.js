import moment from "moment";

const getVisibleExpenses = (expenses, { startDate, endDate, text, sortBy }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatches = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;

      const endDateMatches = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      const textMatches = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatches && endDateMatches && textMatches;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return moment(a.createdAt).isAfter(b.createdAt) ? -1 : 1;
      } else {
        return a.amount > b.amount ? -1 : 1;
      }
    });
};

export default getVisibleExpenses;
