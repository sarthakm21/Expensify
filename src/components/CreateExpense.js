import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

export class CreateExpense extends React.Component {
  onSubmit = (expense) => {
    console.log(expense)
    this.props.addExpense(expense);
    this.props.history.push("/");
  }

  render() {
    return (
      <div>
        <div>
          This is the create element
          <ExpenseForm
            onSubmit={this.onSubmit}
            text="Add Expense"
          />
        </div>
      </div>
    );
  }
}

//aka mapDispatchToProps. Does the same thing to dispatch that mapStateToProps does to state
const mapDispatch = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatch)(CreateExpense);
