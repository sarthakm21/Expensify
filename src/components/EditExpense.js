import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpense extends React.Component {
  onSubmit = (expense) => {
    //Func. gets called by the submit event handler in the form which passes the new data as a parameter
    console.log(expense);
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id); //Can also use props.match.params.id since both are same
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense} //The expense received from mapStateToProps, i.e. the expense with an id same as the props.match.params.id
          onSubmit={this.onSubmit}
          text="Edit Expense"
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(
    (expense) => expense.id === props.match.params.id
  ),
});

const mapDispatch = (dispatch) => ({
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
});

export default connect(mapStateToProps, mapDispatch)(EditExpense);
