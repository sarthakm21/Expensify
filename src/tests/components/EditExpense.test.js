import React from "react";
import { shallow } from "enzyme";
import ExpenseForm from "../../components/ExpenseForm";
import { EditExpense } from "../../components/EditExpense";
import expenses from "../fixtures/expenseData";

let startEditExpense, wrapper, history, startRemoveExpense;
beforeEach(() => {
  startEditExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = {
    push: jest.fn(),
  };
  wrapper = shallow(
    <EditExpense
      startEditExpense={startEditExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={expenses[1]}
    />
  );
});

test("should render edit expense", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle edit expense", () => {
  const testData = expenses[1];
  //Since the onSubmit called by ExpenseForm does not have an id passed.
  const calledWith = {
    amount: testData.amount,
    note: testData.note,
    createdAt: testData.createdAt,
    description: testData.description,
  };
  wrapper.find(ExpenseForm).prop("onSubmit")(calledWith);
  expect(startEditExpense).lastCalledWith(testData.id, calledWith);
  expect(history.push).lastCalledWith("/");
});

test("should handle remove expense", () => {
  wrapper.find("button").simulate("click");
  expect(startRemoveExpense).lastCalledWith(expenses[1].id);
});
