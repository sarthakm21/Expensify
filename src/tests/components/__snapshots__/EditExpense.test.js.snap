// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 10050,
          "createdAt": -345600000,
          "description": "Rent",
          "id": "2",
          "note": "",
        }
      }
      onSubmit={[Function]}
      text="Edit Expense"
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
